Print thread dump on specific signals

Bouncheck opened the pull request:
Adds a signal handler that will print thread dump to standard output on
the following signals: SIGABRT, SIGTERM, SIGINT.
Does not work for SIGKILL or other abrupt shutdowns.

Addresses scylladb/scylla-tools-java/issues/407

dkropachev commented on tools/stress/src/org/apache/cassandra/stress/settings/SettingsLog.java:
```diff
@@ -93,11 +95,12 @@ public static final class Options extends GroupedOptions
         final OptionSimple hdrOutputFile = new OptionSimple("hdrfile=", ".*", null, "Log to a file", false);
         final OptionSimple interval = new OptionSimple("interval=", "[0-9]+(ms|s|)", "1s", "Log progress every <value> seconds or milliseconds", false);
         final OptionSimple level = new OptionSimple("level=", "(minimal|normal|verbose)", "normal", "Logging level (minimal, normal or verbose)", false);
+        final OptionSimple printThreadDumpOnShutdown = new OptionSimple("print-thread-dump-on-shutdown", "", null, "Print local thread dump to the standard output on shutdown (normal, SIGTERM, SIGINT, SIGABRT).", false);
```
I will require additional code on dtest side to understand whethere given c-s support this option or not.
It would be better to just do that by default, but only if `c-s` was terminated.

Bouncheck replied:
Reworked to a version without new option.

dkropachev commented on tools/stress/src/org/apache/cassandra/stress/Stress.java:
```diff
@@ -196,4 +210,20 @@ public void run()
+            System.out.println("Caught signal: " + signal.getName());
+            //Thread dump will be handled by shutdown hook
+        };
+        Signal.handle(new Signal("ABRT"), handler);
```
```suggestion
        Signal.handle(new Signal("ABRT"), handler);
        Signal.handle(new Signal("TERM"), handler);
        Signal.handle(new Signal("INT"), handler);
```

CodeLieutenant replied:
I think there is no need to handle `SIGTERM` and `SIGINT`, since `Runtime.addShutdownHook` already handles them, either we do it through `Signal.handle` or use `Runtime.addShutdownHook`

Bouncheck replied:
As suggested I removed shutdown hook and added those signals to the handler.

CodeLieutenant commented on tools/stress/src/org/apache/cassandra/stress/Stress.java:
```diff
@@ -27,6 +30,9 @@
 import org.apache.cassandra.utils.FBUtilities;
 import org.apache.cassandra.utils.WindowsTimer;
 
+import sun.misc.Signal;
+import sun.misc.SignalHandler;
```
Are the `sun.misc.*` inside the `internal` parts of JDK and shouldn't be accessed?

Bouncheck replied:
Yes, but outside of shutdown hooks I don't know of a way to do this. If it ever gets removed I assume there will be something else for signal handling and in that case this change should be easy to port or ditch altogether since it does not go too deep.

CodeLieutenant replied:
You are right, there is no other way to get to the Unix signals, and handle specific signal through Java Public APIs

dkropachev commented on tools/stress/src/org/apache/cassandra/stress/Stress.java:
```diff
@@ -88,6 +94,14 @@ private static int run(String[] arguments)
 
+            if (settings.log.printThreadDumpOnShutdown) {
+                Runtime.getRuntime().addShutdownHook(new Thread(() -> {
+                    System.out.println("Shutdown thread dump:");
```
It should not print thread dump when it terminates in a regular way, only when it is terminated by signal

Bouncheck replied:
Removed

Bouncheck commented:
Removed shutdown hook. Switched to solution with only signal handler. Now no config options are needed and thread dump will be printed only on three mentioned signals.
